runtime/src/snapshot_utils.rs

```diff
@@ -471,6 +472,107 @@ pub fn move_and_async_delete_path(path: impl AsRef<Path> + Copy) {
+    // This is used to clean up any hardlinks that are no longer referenced by the snapshot dirs.
+
+    let mut account_snapshot_dirs_referenced = HashSet::new();
+    let snapshots = get_bank_snapshots(bank_snapshots_dir);
+    for snapshot in snapshots {
+        let account_hardlinks_dir = snapshot.snapshot_dir.join(SNAPSHOT_ACCOUNTS_HARDLINKS);
+        // loop through entries in the snapshot_hardlink_dir, read the symlinks, add the target to the HashSet
+        fs::read_dir(&account_hardlinks_dir)
+            .unwrap_or_else(|_| {
+                panic!(
+                    "Unable to read snapshot hardlink directory: {}",
+                    account_hardlinks_dir.display()
+                )
+            })
+            .for_each(|entry| {
+                let entry = entry.unwrap_or_else(|_| {
+                    panic!(
+                        "Unable to read snapshot hardlink directory entry: {}",
+                        account_hardlinks_dir.display()
+                    )
+                });
+                let path = entry.path();
+                let target = fs::read_link(&path).unwrap_or_else(|_| {
+                    panic!(
+                        "Unable to read snapshot hardlink directory entry: {}",
+                        path.display()
+                    )
+                });
+                account_snapshot_dirs_referenced.insert(target);
+            });
+    }
+
+    // loop through the account snapshot hardlink directories, if the directory is not in the account_snapshot_dirs_referenced set, delete it
+    for account_snapshot_path in account_snapshot_paths {
+        fs::read_dir(account_snapshot_path)
+            .unwrap_or_else(|_| {
+                panic!(
+                    "Unable to read snapshot hardlink directory: {}",
+                    account_snapshot_path.display()
+                )
+            })
+            .for_each(|entry| {
+                let entry = entry.unwrap_or_else(|_| {
+                    panic!(
+                        "Unable to read snapshot hardlink directory entry: {}",
+                        account_snapshot_path.display()
+                    )
+                });
+                let path = entry.path();
+                if !account_snapshot_dirs_referenced.contains(&path) {
+                    info!(
+                        "Removing orphaned account snapshot hardlink directory: {}",
+                        path.display()
+                    );
+                    move_and_async_delete_path(&path);
+                }
+            });
+    }
+}
+
+/// For all account_paths, set up the run/ and snapshot/ sub directories.
+/// If the sub directories do not exist, the account_path will be cleaned because older version put account files there
+pub fn set_up_account_run_and_snapshot_paths(
+    account_paths: &[PathBuf],
+) -> Result<(Vec<PathBuf>, Vec<PathBuf>)> // (run_paths, snapshot_paths) or error
```

Comment:
The return comment should be part of the doc-comment. That makes it easier for users to see without needing to navigate here. For example, if I mouse hover a use of the function.

```suggestion
/// For all account_paths, set up the run/ and snapshot/ sub directories.
/// If the sub directories do not exist, the account_path will be cleaned because older version put account files there.
/// If successful, returns run and snapshot subdirectory paths: `(run_paths, snapshot_paths)`
pub fn set_up_account_run_and_snapshot_paths(
    account_paths: &[PathBuf],
) -> Result<(Vec<PathBuf>, Vec<PathBuf>)> {
```

Reply:
fixed